test/e2e/traceflow_test.go

```diff
@@ -0,0 +1,330 @@
+							Action:        v1alpha1.Forwarded,
+						},
+						{
+							Component:     v1alpha1.Forwarding,
+							ComponentInfo: "Output",
+							Action:        v1alpha1.Forwarded,
+						},
+					},
+				},
+				{
+					Node: node2,
+					Observations: []v1alpha1.Observation{
+						{
+							Component:     v1alpha1.Forwarding,
+							ComponentInfo: "Classification",
+							Action:        v1alpha1.Received,
+						},
+						{
+							Component:     v1alpha1.Forwarding,
+							ComponentInfo: "Output",
+							Action:        v1alpha1.Delivered,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	localConfig := rest.CopyConfig(data.kubeConfig)
+	client, err := versioned.NewForConfig(localConfig)
+	require.NoError(t, err)
+
+	// Setup 2 NetworkPolicies:
+	// 1. Allow all egress traffic.
+	// 2. Deny ingress traffic on pod with label antrea-e2e = node1Pods[1]. So flow node1Pods[0] -> node1Pods[1] will be dropped.
+	var allowAllEgress *networkingv1.NetworkPolicy
+	if allowAllEgress, err = data.createNPAllowAllEgress("test-networkpolicy-allow-all-egress"); err != nil {
+		t.Fatalf("Error when creating network policy: %v", err)
+	}
+	defer func() {
+		if err = data.deleteNetworkpolicy(allowAllEgress); err != nil {
+			t.Fatalf("Error when deleting network policy: %v", err)
+		}
+	}()
+
+	var denyIngress *networkingv1.NetworkPolicy
+	if denyIngress, err = data.createNPDenyIngress("antrea-e2e", node1Pods[1], "test-networkpolicy-deny-ingress"); err != nil {
+		t.Fatalf("Error when creating network policy: %v", err)
+	}
+	defer func() {
+		if err = data.deleteNetworkpolicy(denyIngress); err != nil {
+			t.Fatalf("Error when deleting network policy: %v", err)
+		}
+	}()
+
+	// Create 2 traceflows:
+	// 1. node1Pods[0] -> node1Pods[1], intra node1.
+	// 2. node1Pods[0] -> node2Pods[0], inter node1 and node2.
+	if tfIntraNode, err = client.OpsV1alpha1().Traceflows().Create(context.TODO(), tfIntraNode, v1.CreateOptions{}); err != nil {
+		t.Fatalf("Error when creating traceflow: %v", err)
+	}
+	defer func() {
+		if err := client.OpsV1alpha1().Traceflows().Delete(context.TODO(), tfIntraNode.Name, v1.DeleteOptions{}); err != nil {
+			t.Fatalf("Error when deleting traceflow: %v", err)
+		}
+	}()
+
+	if tfInterNode, err = client.OpsV1alpha1().Traceflows().Create(context.TODO(), tfInterNode, v1.CreateOptions{}); err != nil {
+		t.Fatalf("Error when creating traceflow: %v", err)
+	}
+	defer func() {
+		if err := client.OpsV1alpha1().Traceflows().Delete(context.TODO(), tfInterNode.Name, v1.DeleteOptions{}); err != nil {
+			t.Fatalf("Error when deleting traceflow: %v", err)
+		}
+	}()
+
+	// Wait for traceflows to finish.
+	if err = wait.Poll(1*time.Second, traceflowTimeout, func() (bool, error) {
+		done := 0
+		for i := 0; i < len(testcases); i++ {
+			tc := testcases[i]
+			tf := tc.tf
+			if tf, err = client.OpsV1alpha1().Traceflows().Get(context.TODO(), tf.Name, v1.GetOptions{}); err != nil || tf.Status.Phase != tc.expectedPhase {
```

Comment:
Why not check all fields here but start another loop at L203?
You may first check phase
if it doesn't match, return `false, nil`;
if it matches, check other fields, if they don't match, return `false, fmt.Errorf("expected xxx, got xxx")`, the poll will terminate and return error;
if all fields match, done++;

Reply:
Updated.